lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -355,31 +362,33 @@ export default class BitcoinProcessor {
 
   private async getStartingBlockForInitialization (): Promise<IBlockInfo> {
 
-    // Assume that we're going to start form the genesis block
-    let startingBlock = this.genesisBlockNumber;
+    // Look in the transaction store to figure out the last block that we need to
+    // start from.
+    const lastSavedTransaction = (await this.transactionStore.getLastTransaction());
 
-    // We look at the latest data that we saved in the quantile calculator. The latest
-    // Id in there tells us which group of blocks have been processed.
-    const lastSavedGroupId = await this.quantileCalculator.getLastGroupId();
+    // If there's nothing saved in the DB then let's start from the genesis block
+    if (!lastSavedTransaction) {
+      return this.bitcoinClient.getBlockInfoFromHeight(this.genesisBlockNumber);
+    }
 
-    if (lastSavedGroupId) {
-      // Then we need to start from the block corresponding to that groupId
-      startingBlock = this.getStartingBlockFromGroupId(lastSavedGroupId);
+    // If we are here then it means that there is a potential starting point in the DB.
+    // Since we are initializing, it is quite possible that the last block that we processed
+    // (and saved in the db) has been forked. Check for the fork.
+    const lastSavedBlockIsValid = await this.verifyBlock(lastSavedTransaction.transactionTime, lastSavedTransaction.transactionTimeHash);
 
-      // We want to delete the last group Id because we might not have saved all the transaction(s)
-      // in that group. So removing that ensures that we redo this block which will ensure that
-      // all the transactions are processed again.
-      await this.quantileCalculator.removeGroupsGreaterThanOrEqual(lastSavedGroupId);
-    }
+    let lastValidBlock: IBlockInfo;
 
-    // Remove everything after the starting block as we are (re)starting from that point
-    const startingBlockFirstTxnNumber = TransactionNumber.construct(startingBlock, 0);
-    await this.transactionStore.removeTransactionsLaterThan(startingBlockFirstTxnNumber - 1);
+    if (lastSavedBlockIsValid) {
+      // There was no fork ... let's put the system DBs in the correct state.
+      lastValidBlock = await this.revertQuantileAndTransactionDbsTo(lastSavedTransaction.transactionTime);
+    } else {
+      // There was a fork so we need to revert. The revert function peforms all the correct
+      // operations and puts the system in the correct state and returns the last valid block.
+      lastValidBlock = await this.revertBlockchainCache();
+    }
 
-    return {
-      height: startingBlock,
-      hash: await this.bitcoinClient.getBlockHash(startingBlock)
-    };
+    // Our starting block is the one after the last-valid-block
+    return this.bitcoinClient.getBlockInfoFromHeight(lastValidBlock.height + 1);
   }
```

Comment:
Sanity: this section looks very similar to `getStartingBlockForPeriodicPoll`, no point in consolidating?

Reply:
The two functions are different enough that we can't really merge them. I'll keep them as they are.